dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java

```diff
@@ -216,38 +208,90 @@ private DependencyNode resolveCompileTimeDependencies(
    * version seen. This is how Maven normally operates.
    */
-  public DependencyGraphResult buildGraph(Dependency dependency)
-      throws RepositoryException {
-    // root node
-    DependencyNode node = resolveCompileTimeDependencies(new DefaultDependencyNode(dependency));
-    return levelOrder(node);
+  public DependencyGraphResult buildGraph(Dependency dependency) {
+    return buildDependencyGraph(
+        ImmutableList.of(new DefaultDependencyNode(dependency)), GraphTraversalOption.NONE);
+  }
+
+  private DependencyGraphResult buildDependencyGraph(
+      List<DependencyNode> dependencyNodes, GraphTraversalOption traversalOption) {
+    boolean includeProvidedScope =
+        traversalOption == GraphTraversalOption.FULL_DEPENDENCY_WITH_PROVIDED;
+    DependencyNode node;
+    ImmutableSet.Builder<UnresolvableArtifactProblem> artifactProblems = ImmutableSet.builder();
+
+    try {
+      node = resolveDependencyGraph(dependencyNodes, includeProvidedScope);
+    } catch (DependencyResolutionException ex) {
+      DependencyResult result = ex.getResult();
+      node = result.getRoot();
+      for (ArtifactResult artifactResult : result.getArtifactResults()) {
+        Artifact resolvedArtifact = artifactResult.getArtifact();
+        if (resolvedArtifact != null) {
+          continue;
+        }
+        Artifact requestedArtifact = artifactResult.getRequest().getArtifact();
+        artifactProblems.add(createUnresolvableArtifactProblem(node, requestedArtifact));
+      }
+    }
+
+    DependencyGraphResult result = levelOrder(node, traversalOption);
+    // Duplicate problems found in resolveDependencyGraph and levelOrder are removed by ImmutableSet
+    artifactProblems.addAll(result.getArtifactProblems());
+
+    return new DependencyGraphResult(result.getDependencyGraph(), artifactProblems.build());
+  }
+
+  /**
+   * Returns a problem describing that {@code artifact} is unresolvable in the {@code
+   * dependencyGraph}.
+   */
+  public static UnresolvableArtifactProblem createUnresolvableArtifactProblem(
+      DependencyNode dependencyGraph, Artifact artifact) {
+    ImmutableList<List<DependencyNode>> paths = findArtifactPaths(dependencyGraph, artifact);
+    if (paths.isEmpty()) {
+      // On certain conditions, Maven throws ArtifactDescriptorException even when the
+      // (transformed) dependency dependencyGraph does not contain the problematic artifact any
+      // more.
+      // https://issues.apache.org/jira/browse/MNG-6732
+      return new UnresolvableArtifactProblem(artifact);
+    } else {
+      return new UnresolvableArtifactProblem(paths.get(0));
+    }
+  }
+
+  private static ImmutableList<List<DependencyNode>> findArtifactPaths(
+      DependencyNode root, Artifact artifact) {
+    String coordinates = Artifacts.toCoordinates(artifact);
+    DependencyFilter filter =
+        (node, parents) ->
+            node.getArtifact() != null
```

Comment:
artifact is null at a root dummy node.

Reply:
This would be more useful as a code comment than a review comment.

Reply:
Added source code comment.